File: ESP32/Queue/QueueManager.cpp
#include "QueueManager.h"
#include <algorithm>

QueueManager::QueueManager(size_t maxSize, size_t numberOfLines)
    : m_maxSize(maxSize), m_numberOfLines(numberOfLines), m_totalPeople(0) {
    for (int i = 1; i <= m_numberOfLines; ++i) {
        auto line = make_shared<Line>(Line{i, 0});
        lineMap[i] = line;
        queue.insert(line);
    }
}

bool QueueManager::enqueue() {
    if (isFull()) {
        return false;
    }

    if (queue.empty()) {
        return false;
    }

    auto it = queue.begin();
    auto line = *it;
    queue.erase(it);
    line->peopleCount++;
    queue.insert(line);
    ++m_totalPeople;

    return true;
}

bool QueueManager::dequeue(int lineNumber) {
    auto it = lineMap.find(lineNumber);
    if (it == lineMap.end() || it->second->peopleCount == 0) {
        return false; // Invalid line number or no people in line
    }

    auto line = it->second;
    auto setIt = queue.find(line);
    if (setIt != queue.end()) {
        queue.erase(setIt);
    }

    line->peopleCount--;
    queue.insert(line);
    --m_totalPeople;

    return true;
}

bool QueueManager::enqueueOnLine(int lineNumber) {
    if (isFull()) {
        return false;
    }

    auto it = lineMap.find(lineNumber);
    if (it == lineMap.end()) {
        return false;
    }

    auto line = it->second;
    auto setIt = queue.find(line);
    if (setIt != queue.end()) {
        queue.erase(setIt);
    }

    line->peopleCount++;
    queue.insert(line);
    ++m_totalPeople;

    return true;
}

size_t QueueManager::size() const {
    return m_totalPeople;
}

bool QueueManager::isEmpty() const {
    return m_totalPeople == 0;
}

bool QueueManager::isFull() const {
    if (m_maxSize == 0) {
        return false;
    }
    return m_totalPeople >= m_maxSize;
}

int QueueManager::getNextLineNumber() const {
    if (queue.empty()) {
        return -1;
    }
    return (*queue.begin())->lineNumber;
}

size_t QueueManager::getNumberOfLines() const {
    return m_numberOfLines;
}

int QueueManager::getLineCount(int lineNumber) const {
    auto it = lineMap.find(lineNumber);
    if (it == lineMap.end()) {
        return -1;
    }
    return it->second->peopleCount;
}

void QueueManager::setLineCount(int lineNumber, int count) {
    auto it = lineMap.find(lineNumber);
    if (it == lineMap.end()) {
        return;
    }

    auto line = it->second;
    auto setIt = queue.find(line);
    if (setIt != queue.end()) {
        queue.erase(setIt);
    }

    m_totalPeople -= line->peopleCount;
    line->peopleCount = max(0, count);
    m_totalPeople += line->peopleCount;
    queue.insert(line);
}

void QueueManager::reset() {
    for (auto &entry : lineMap) {
        auto line = entry.second;
        auto setIt = queue.find(line);
        if (setIt != queue.end()) {
            queue.erase(setIt);
        }
        m_totalPeople -= line->peopleCount;
        line->peopleCount = 0;
        queue.insert(line);
    }
    m_totalPeople = 0;
}

File: flutter_app/lib/controllers/queue_manager.dart
/// Dart adaptation of the C++ QueueManager for simulation and UI decisions.
/// Chooses the line with the fewest people (ties -> lowest line number).
class QueueManagerDart {
  QueueManagerDart({required int maxSize, required int numberOfLines})
      : _maxSize = maxSize,
        _numberOfLines = numberOfLines {
    for (var i = 1; i <= _numberOfLines; i++) {
      _lines[i] = 0; // peopleCount
    }
  }

  final int _maxSize; // 0 means unlimited
  final int _numberOfLines;
  int _totalPeople = 0;
  final Map<int, int> _lines = {}; // lineNumber -> peopleCount

  bool get isEmpty => _totalPeople == 0;
  bool get isFull => _maxSize != 0 && _totalPeople >= _maxSize;
  int get totalPeople => _totalPeople;
  int get numberOfLines => _numberOfLines;

  int getLineCount(int lineNumber) => _lines[lineNumber] ?? -1;

  bool enqueue() {
    if (isFull) return false;
    final ln = getNextLineNumber();
    if (ln == -1) return false;
    _lines[ln] = _lines[ln]! + 1;
    _totalPeople++;
    return true;
  }

  bool enqueueOnLine(int lineNumber) {
    if (isFull) return false;
    if (!_lines.containsKey(lineNumber)) return false;
    _lines[lineNumber] = _lines[lineNumber]! + 1;
    _totalPeople++;
    return true;
  }

  bool dequeue(int lineNumber) {
    if (!_lines.containsKey(lineNumber)) return false;
    if (_lines[lineNumber]! == 0) return false;
    _lines[lineNumber] = _lines[lineNumber]! - 1;
    _totalPeople--;
    return true;
  }

  void setLineCount(int lineNumber, int count) {
    if (!_lines.containsKey(lineNumber)) return;
    _totalPeople -= _lines[lineNumber]!;
    _lines[lineNumber] = count < 0 ? 0 : count;
    _totalPeople += _lines[lineNumber]!;
  }

  void reset() {
    _lines.updateAll((key, value) => 0);
    _totalPeople = 0;
  }

  int getNextLineNumber() {
    if (_lines.isEmpty) return -1;
    int bestLine = -1;
    int bestCount = 1 << 30;
    _lines.forEach((line, count) {
      if (count < bestCount || (count == bestCount && line < bestLine)) {
        bestLine = line;
        bestCount = count;
      }
    });
    return bestLine;
  }

  Map<String, dynamic> toJson() => {
        'lines': _lines.map((k, v) => MapEntry(k.toString(), v)),
        'totalPeople': _totalPeople,
        'recommendedLine': getNextLineNumber(),
      };
}
